test/utils_test.dart

```diff
@@ -318,13 +318,35 @@ strings:
     test'''));
       });
 
+      test('generates folded strings with trailing line break and space', () {
+        final doc = YamlEditor('');
+
+        doc.update(
+          [],
+          wrapAsYamlNode('test\ntest\n ', scalarStyle: ScalarStyle.FOLDED),
+        );
+
+        expect(doc.toString(), equals('>+\n  test\n\n  test\n   '));
+      });
+
       test('generates literal strings properly', () {
         final doc = YamlEditor('');
         doc.update(
             [], wrapAsYamlNode('test\ntest', scalarStyle: ScalarStyle.LITERAL));
         expect(doc.toString(), equals('|-\n  test\n  test'));
       });
 
+      test('generates literal string with trailing line break and space', () {
+        final doc = YamlEditor('');
+
+        doc.update(
+          [],
+          wrapAsYamlNode('test\ntest\n ', scalarStyle: ScalarStyle.LITERAL),
+        );
+
+        expect(doc.toString(), equals('|+\n  test\n  test\n   '));
+      });
```

Comment:
So I tried something like this in an online editor that converted yaml to JSON and I kept getting the resulting string as: `"test\ntest\n \n"`

Probably because the editor always ended the YAML document with a line-break: `|+\n  test\n  test\n   ` would become: `|+\n  test\n  test\n   \n` (because files normally end with a line break).

I tried to recreate it, and it wasn't a problem. But I'm guessing it's unwise for us to produce a YAML document where adding empty lines to the YAML document causes it to change its semantics.

If I parse the YAML document `|+\n  test\n  test\n   \n`, then I won't get `"test\ntest\n \n"`.

Example:
 * YAML: `A: |+\n    test\n    test\n     `
   * YAML to JSON: `{"A":"test\ntest\n "}` (this is correct!)
 * If I add `\n` at the end of the YAML I get: `A: |+\n    test\n    test\n     \n`
   * YAML to JSON: `{"A":"test\ntest\n \n"}` (this is wrong)
 * If I add `\nB: true` at the end of the YAML I get: `A: |+\n    test\n    test\n     \nB: true`
  * YAML to JSON: `{"A":"test\ntest\n \n","B":true}` (this is wrong)

So if we allow this as a literal string, we are VERY likely to cause all sorts of breakages next time we try to append a key.

Reply:
I did another variant of this test case:
```dart
      test('generates literal string with trailing line break and space (2)',
          () {
        final doc = YamlEditor('key1: true');

        doc.update(
          ['key2'],
          wrapAsYamlNode('test\ntest\n ', scalarStyle: ScalarStyle.LITERAL),
        );
      });
```
it causes a crash:
```
  Assertion failed: (package:yaml_edit) Modification did not result in expected result.
  
  # YAML before edit:
  > key1: true
  
  # YAML after edit:
  > key1: true
  > key2: |+
  >     test
  >     test
  >      
  > 
```

------------------------

I'm beginning to see how a literal string ending in `\n ` might have to be encoded as a double quoted string. :rofl:

Reply:
Should we consider falling back to double quoted mode for literal strings ending in newline + whitespace?

The whole adding a new key afterwards seems like a reasonable thing, and it's a little sad that this can't work.



Of course, we can't know what document the user gives us. So we could also view this as an orthogonal bug to do with adding keys to block maps and entries to block lists, because those are the scenarios where this could go wrong.

I don't know if there is a reasonably way to fix adding keys to block maps that have a preceeding key-value pair where the value is a literal string ending in `\n ` :rofl:  -- definitely we should not try to fix that in this PR.

Reply:
Also isn't it likely that the chomping indicator `+` is always going to cause us this kind of problems? Just asking, I'm not sure.

Reply:
@jonasfj The final `\n` just before the next key is the issue. 

It seems chomping indicators only control the final line break before going to the next `YamlScalar` (value in list or key in map) and not content within the scalar itself (which is  only controlled by either `>` or `|`)

After some investigation, we can exclusively use the strip chomping indicator `-` for both `ScalarStyle.FOLDED` and `ScalarStyle.LITERAL`.  It always indicates we don't want the last line break that's just before the next key in map or value in the list.

😅 The fixes I've tried coming up with for the `keep chomping indicator +` bug were all shots in the dark and just another bug creeping in. 

I'm sorry and thanks! The bug was staring right at us from the beginning. This is the line I should have removed before fixing `_tryYamlEncodeFolded`

https://github.com/dart-lang/yaml_edit/blob/f3feb50dab3b36fe5ec115e78667712473e2e882/lib/src/strings.dart#L89-L93